// @webex/webex-core/dist/lib/services/service-url.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _ampersandState = require('ampersand-state');

var _ampersandState2 = _interopRequireDefault(_ampersandState);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-underscore-dangle */
/**
 * @class
 */
var ServiceUrl = _ampersandState2.default.extend({
  namespace: 'ServiceUrl',

  props: {
    defaultUrl: ['string', true, undefined],
    hosts: ['array', false, function () {
      return [];
    }],
    name: ['string', true, undefined]
  },

  /**
   * Generate a host url based on the host
   * uri provided.
   * @param {string} hostUri
   * @returns {string}
   */
  _generateHostUrl: function _generateHostUrl(hostUri) {
    var url = _url2.default.parse(this.defaultUrl);

    // setting url.hostname will not apply during Url.format(), set host via
    // a string literal instead.
    url.host = '' + hostUri + (url.port ? ':' + url.port : '');

    return _url2.default.format(url);
  },


  /**
   * Generate a list of urls based on this
   * `ServiceUrl`'s known hosts.
   * @returns {string[]}
   */
  _getHostUrls: function _getHostUrls() {
    var _this = this;

    return this.hosts.map(function (host) {
      return {
        url: _this._generateHostUrl(host.host),
        priority: host.priority
      };
    });
  },


  /**
   * Get the current host url with the highest
   * priority.
   * @returns {string}
   */
  _getPriorityHostUrl: function _getPriorityHostUrl() {
    if (this.hosts.length === 0) {
      return this.defaultUrl;
    }

    return this._generateHostUrl(this.hosts.reduce(function (previous, current) {
      return previous.priority > current.priority || !previous.homeCluster ? current : previous;
    }).host);
  },


  /**
   * Attempt to remove a host from this `ServiceUrl`
   * and return true if the provided `url` had a host
   * that could be successfully removed or if the url
   * is assocaited with a defaultUrl, and cannot be
   * removed.
   * @param {string} url
   * @returns {boolean}
   */
  removeHost: function removeHost(url) {
    if (url === this.defaultUrl) {
      return true;
    }

    var _Url$parse = _url2.default.parse(url),
        hostname = _Url$parse.hostname;

    var foundHost = this.hosts.find(function (hostObj) {
      return hostObj.host === hostname;
    });

    this.hosts.splice(this.hosts.indexOf(foundHost), 1);

    return foundHost !== undefined;
  },


  /**
   * Get the current `defaultUrl` or generate a url
   * using the host with the highest priority via host rendering.
   * @param {boolean} [priorityHost]
   * @returns {string}
   */
  get: function get(priorityHost) {
    if (!priorityHost) {
      return this.defaultUrl;
    }

    return this._getPriorityHostUrl();
  }
});
/* eslint-enable no-underscore-dangle */

exports.default = ServiceUrl;
